### DISCUSSION TITLE
[Chapter 08] GenerationType 종류 필기

### DISCUSSION THREAD

**Discussion by Youngju-Jang** [cowaine/spring-boot-for-msa#81] [Chapter 08. Data Persistence] [answered]
## 1. GenerationType.TABLE
- TBl_SEQUENCES 채번 테이블 : 공통테이블

```JAVA
 @GeneratedValue(strategy = GenerationType.TABLE, generator = "sequenceGenerator") // 1)
 @TableGenerator(
      name = "sequenceGenerator",  // 1) 에서 사용할 이름
      table = "TBL_SEQUENCES", // 채번용 공통테이블
      pkColumnName = "SEQUENCE_NAME",  // TBL_SEQUENCES 의 기본키
      pkColumnValue = "HOTEL_SEQUENCE", // 채번테이블 기본키에 저장할 키 값. SEQUENCE_NAME 필드에 저장됨
      valueColumnName = "SEQUENCE_VALUE", // 채번테이블 채번값 조회, 업데이트할 필드이름
      initialValue = 10000, allocationSize = 100 // 초기값, 증가값
 )
 @Column
 private Long hotelId;
 ```

 ## 2. GenerationType.SEQUENCE
- DB에서 제공하는 시퀀스기능을 사용하여 기본키 값 설정

// 1에서 시작, 1씩증가하는 sequence정의 ddl
CREATE SEQUENCE hotel_sequence MINVALUE 1 START WITH 1 INCREMENT BY 1;

```JAVA
@SequenceGenerator(name = "hotel_generator", sequenceName = "hotel_sequence")
@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "hotel_generator")
@Column
private Long hotelId;
```

## GenerationType.IDENTITY
- 디비에서 제공하는 자동증가기능으로 기본 키 값을 설정

## GenerationType.AUTO
- 디비에 따라 적절한 키생서전략 매핑
    - 오라클 : 시퀀스방식,
    - mysql: IDENTITY방식
```java
@Id
@GeneratedValue(strategy = GenerationType.AUTO) // <<<
@Column
private Long hotelId;
```

- 저같은 경우 IDENTITY만 써봤던 것 같습니다
- 얼핏 어떤 케이스에서 오류가 났던 기억이 있는데, 제일 처음 테이블생성 후 초기값 설정이 안되어서 난 오류정도로 기억합니다 만, 자세히는 기억이 안나는군요..

**Comment by coalong** [comment, marked as answer]
> 얼핏 어떤 케이스에서 오류가 났던 기억이 있는데, 제일 처음 테이블생성 후 초기값 설정이 안되어서 난 오류정도로 기억합니다 만, 자세히는 기억이 안나는군요..


저도 예전에 프로젝트 할 때, 
Postgresql 을 사용하면서 GenerationType.IDENTITY 을 사용했다가 오류가 났던 적이 있었습니다.
GenerationType.SEQUENCE 로 수정해서 해결한 경험이 있는데 자세한 원인을 기록하지 않은 채로 넘어가서 기억이 안나네요...
(기록이 중요하다는 걸 다시 생각했습니다..)

아래 내용과 관련있는 것 같은데 다시 예시코드를 작성하면서 좀 더 알아봐도 좋을 것 같습니다.


[postgresql에서-jpa-자동-키-생성-오류](https://sejoung.github.io/2021/09/2021-09-16-spring_data_jpa_error_postgresql/#postgresql%EC%97%90%EC%84%9C-jpa-%EC%9E%90%EB%8F%99-%ED%82%A4-%EC%83%9D%EC%84%B1-%EC%98%A4%EB%A5%98)
[Hibernate User Guide](https://docs.jboss.org/hibernate/orm/5.2/userguide/html_single/Hibernate_User_Guide.html#identifiers-generators-identity)
[PostgreSQL SERIAL or IDENTITY column and Hibernate IDENTITY generator](https://vladmihalcea.com/postgresql-serial-column-hibernate-identity/)

**Comment by Irisation23** [comment]
### 질의 응답 이외의 CheckPoint

1. 차후 예외가 발견 된다면 예외 내용을 같이 첨부 하는것이 좋아보임.
